server/src/main/java/org/elasticsearch/indices/IndexDescriptor.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.indices;
+
+import org.apache.lucene.util.automaton.Automaton;
+import org.apache.lucene.util.automaton.CharacterRunAutomaton;
+import org.apache.lucene.util.automaton.RegExp;
+import org.elasticsearch.cluster.metadata.Metadata;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * An "associated index" is an index that is related to or derived from a system
+ * index, but should not be considered a system index, usually because it is
+ * meant to be visible to users. However, if it goes out of sync with its
+ * related system indices, it may become unreliable or useless. Hence, when
+ * taking a snapshot of a feature, we want the associated index to be included
+ * in the snapshot, and, likewise, we want an associated index to be restored
+ * when a feature snapshot is restored.
+ */
+public class IndexDescriptor {
+    /** A pattern, either with a wildcard or simple regex.*/
+    protected final String indexPattern;
+
+    /** A description of the index or indices */
+    protected final String description;
+
+    /** Used to determine whether an index name matches the {@link #indexPattern} */
+    private final CharacterRunAutomaton indexPatternAutomaton;
+
+    /**
+     * Create a descriptor for an index associated with a feature
+     * @param indexPattern Pattern for associated indices, which can include wildcards
+     *                     or regex-like character classes
+     * @param description A short description of the purpose of this index
+     */
+    public IndexDescriptor(String indexPattern, String description) {
+        Objects.requireNonNull(indexPattern, "associated index pattern must not be null");
```

Comment:
These error messages should say `system index` if a system index throws the error.